trino-s3-proxy/src/main/java/io/trino/s3/proxy/server/TrinoS3ProxyModuleBuilder.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.s3.proxy.server;
+
+import com.google.inject.Module;
+import com.google.inject.binder.LinkedBindingBuilder;
+import io.trino.s3.proxy.server.credentials.AssumedRoleProvider;
+import io.trino.s3.proxy.server.credentials.CredentialsProvider;
+
+import java.util.function.Consumer;
+
+import static com.google.inject.multibindings.OptionalBinder.newOptionalBinder;
+import static java.util.Objects.requireNonNull;
+
+public interface TrinoS3ProxyModuleBuilder
+{
+    static Builder builder()
+    {
+        return new Builder();
+    }
+
+    final class Builder
+    {
+        private Consumer<LinkedBindingBuilder<CredentialsProvider>> credentialsProviderBinder;
+        private Consumer<LinkedBindingBuilder<AssumedRoleProvider>> assumedRoleProviderBinder;
+
+        private Builder() {}
+
+        public Builder withCredentialsProvider(Consumer<LinkedBindingBuilder<CredentialsProvider>> credentialsProviderBinder)
+        {
+            this.credentialsProviderBinder = requireNonNull(credentialsProviderBinder, "credentialsProviderBinder is null");
+            return this;
+        }
+
+        public Builder withAssumedRoleProvider(Consumer<LinkedBindingBuilder<AssumedRoleProvider>> assumedRoleProviderBinder)
+        {
+            this.assumedRoleProviderBinder = requireNonNull(assumedRoleProviderBinder, "assumedRoleProviderBinder is null");
+            return this;
+        }
+
+        public Module build()
+        {
+            return binder -> {
+                if (credentialsProviderBinder != null) {
```

Comment:
Would this be nicer with optionals? The type would be quite a mouthful though :)

Reply:
Normally, I do Optional everywhere. Sure.